staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -545,3 +573,106 @@ func (a *Authenticator) wrapCmdRunErrorLocked(err error) error {
+	case api.PluginPolicyAllowlist:
+		return a.checkAllowlist()
+	default:
+		return fmt.Errorf("unknown plugin policy %q", a.execPluginPolicy.PolicyType)
+	}
+}
+
+func (a *Authenticator) checkAllowlist() error {
+	pluginAbsPath, err := exec.LookPath(a.cmd)
+	if err != nil {
+		return fmt.Errorf("could not resolve path for plugin %q: %w", a.cmd, err)
+	}
+
+	errs := make([]error, 0, len(a.execPluginPolicy.Allowlist))
+	for _, entry := range a.execPluginPolicy.Allowlist {
+		err := itemGreenlights(&entry, pluginAbsPath)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		return nil // an item in the allowlist matched, return nil to indicate success
+	}
+
+	return fmt.Errorf("%q is not permitted by the credential plugin allowlist\n%w", pluginAbsPath, utilerrors.NewAggregate(errs))
+}
+
+var emptyAllowlistEntry api.AllowlistEntry
+
+// alEntry MUST be non-nil
+func itemGreenlights(alEntry *api.AllowlistEntry, pluginAbsPath string) error {
+	// if no fields are specified, this is a user error. To avoid fail-open
+	// behavior, an empty entry must not allow anything.
+	if *alEntry == emptyAllowlistEntry {
+		return fmt.Errorf("allowlist entry is empty")
+	}
+
+	if entryName := alEntry.Name; len(entryName) > 0 {
+		entryAbsPath, err := exec.LookPath(entryName)
+		if err != nil || pluginAbsPath != entryAbsPath {
+			return fmt.Errorf("allowlist entry %q is not a match for %q: %w", entryName, pluginAbsPath, err)
+		}
+	}
+
+	return nil
+}
+
+func isEmptyPolicy(policy api.PluginPolicy) bool {
+	return len(policy.PolicyType) == 0 && policy.Allowlist == nil
+}
+
+func ValidatePluginPolicy(policy api.PluginPolicy) error {
+	if len(policy.PolicyType) == 0 {
+		return fmt.Errorf("unspecified plugin policy")
+	}
+
+	switch policy.PolicyType {
+	case api.PluginPolicyAllowAll, api.PluginPolicyDenyAll:
+		if policy.Allowlist != nil {
+			return fmt.Errorf("misconfigured credential plugin allowlist: plugin policy is %q but allowlist is non-nil", policy.PolicyType)
+		}
+		return nil
+	case api.PluginPolicyAllowlist:
+		return validateAllowlist(policy.Allowlist)
+	default:
+		return fmt.Errorf("unknown plugin policy: %q", policy.PolicyType)
+	}
+}
+
+func validateAllowlist(list []api.AllowlistEntry) error {
+	// This will be the case if the user has misspelled the field name for the
+	// allowlist. Because this is a security knob, fail immediately rather than
+	// proceed when the user has made a mistake.
+	if list == nil {
+		return fmt.Errorf("credential plugin policy set to %q, but allowlist is unspecified", api.PluginPolicyAllowlist)
+	}
+
+	if len(list) == 0 {
+		return fmt.Errorf("credential plugin policy set to %q, but allowlist is empty; use %q policy instead", api.PluginPolicyAllowlist, api.PluginPolicyDenyAll)
+	}
+
+	for i, item := range list {
+		if item == emptyAllowlistEntry {
+			return fmt.Errorf("misconfigured credential plugin allowlist: empty allowlist entry #%d", i+1)
+		}
```

Comment:
should we make sure the name fields are normalized (https://pkg.go.dev/path/filepath#Clean is a no-op)?